### PR TITLE
docs(NODE-6456): document CSOT pt 1

### PR DESCRIPTION
**Pull request by W-A-James**
#### What is changing?
* Add API docs to `timeoutMS` and `timeoutMode` options
* Add API docs to `CursorTimeoutMode`
* Make `timeoutMS` and `timeoutMode` public

### Double check the following

- [X] Ran `npm run check:lint` script
- [X] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [X] PR title follows the [correct format](https://www.conventionalcommits.org/en/v1.0.0/): `type(NODE-xxxx)[!]: description`
  - Example: `feat(NODE-1234)!: rewriting everything in coffeescript`
- [ ] Changes are covered by tests
- [ ] New TODOs have a related JIRA ticket

### CODE REVIEW

**Comment by W-A-James** [comment]
Question: Did we want to mark the CSOT options as public or as experimental for the release?

**Review by nbbeeken — APPROVED** [review]
LGTM!

I just fixed a bit of formatting: public tags on a new line, example tags directly proceed code snippets.

I also fixed our timeoutMode enum not exported from the top-level index file, and added "type" exports to 'AnyBulkWriteOperation', 'BulkWriteOptions' which fix an outstanding erroneous issue where when loaded with ts-node those appear as "undefined" exports.

The diff on the EXPECTED_EXPORTS may look strange, I alphabetized it to make 👁️ eye inspection a bit easier.

File: src/operations/operation.ts [hunk 1]
```diff
@@ -61,9 +61,9 @@ export abstract class AbstractOperation<TResult = any> {
 
   options: OperationOptions;
 
-  /** @internal */
+  /** TODO(NODE-6197): remove dead code */
```
Comment (aditi-khare-mongoDB) [inline src/operations/operation.ts:64]:
```suggestion
  /** @internal TODO(NODE-6197): remove dead code */
```

File: test/unit/index.test.ts [hunk 2]
```diff
@@ -5,14 +5,7 @@ import { expect } from 'chai';
 import * as mongodb from '../../src/index';
 import { setDifference } from '../mongodb';
 
-/**
```
Comment (aditi-khare-mongoDB) [inline test/unit/index.test.ts:8]: what's the motivation for the changes in this file?

**Review by aditi-khare-mongoDB — CHANGES_REQUESTED** [review]

**Comment by aditi-khare-mongoDB** [comment]
Also, let's add in the `@experimental` tag in instead of `@public`

File: src/cursor/abstract_cursor.ts [hunk 3]
```diff
@@ -116,9 +143,33 @@ export interface AbstractCursorOptions extends BSONSerializeOptions {
...
+   * ```ts
+   * const cursor = collection.find({}, {timeoutMS: 100, timeoutMode: 'iteration'});
+   * for await (const doc of cursor) {
+   *  // process doc
+   *  // This will throw a timeout error if any of the iterator's `next()` calls takes more than 100ms, but
+   *  // will continue to iterate successfully otherwise, regardless of the number of batches.
+   * }
+   * ```
+   *
+   * @example
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/cursor/abstract_cursor.ts]: I sync-ed it up, but the headers are what make the text to big so those are left out (I think example tags are supposed to be used right before code snippets anyway)

File: src/cursor/abstract_cursor.ts [hunk 4]
```diff
@@ -61,15 +61,42 @@ export interface CursorStreamOptions {
...
+ * ```ts
+ * const cursor = collection.find({}, {timeoutMS: 100, timeoutMode: 'iteration'});
+ * for await (const doc of cursor) {
+ *  // process doc
+ *  // This will throw a timeout error if any of the iterator's `next()` calls takes more than 100ms, but
+ *  // will continue to iterate successfully otherwise, regardless of the number of batches.
+ * }
+ * ```
+ *
+ * # Example showing use of `'cursorLifetime'`
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/cursor/abstract_cursor.ts]: applied

File: src/cursor/run_command_cursor.ts [hunk 5]
```diff
@@ -19,9 +19,40 @@ import {
...
+   * will continue to iterate successfully otherwise, regardless of the number of batches.
+   * }
+   * ```
+   *
+   * @example
+   * ```ts
+   * const cursor = collection.find({}, { timeoutMS: 1000, timeoutMode: 'cursorLifetime' });
+   * const docs = await cursor.toArray(); // This entire line will throw a timeout error if all batches are not fetched and returned within 1000ms.
+   * ```
+   */
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/cursor/run_command_cursor.ts:56]: fixed

File: src/db.ts [hunk 6]
```diff
@@ -222,7 +222,6 @@ export class Db {
     return this.s.namespace.toString();
   }
 
-  /** @internal */
   get timeoutMS(): number | undefined {
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/db.ts]: Nope, fixed

File: src/operations/operation.ts (hunk 1 above)
Comment (nbbeeken) [reply src/operations/operation.ts]: done

File: test/unit/index.test.ts (hunk 2 above)
Comment (nbbeeken) [reply test/unit/index.test.ts:8]: https://github.com/mongodb/node-mongodb-native/pull/4292#pullrequestreview-2408002618

I kinda covered it in here, let me know if I can clarify something.

File: src/cursor/abstract_cursor.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (aditi-khare-mongoDB) [reply src/cursor/abstract_cursor.ts]: Sounds good!

**Review by aditi-khare-mongoDB — CHANGES_REQUESTED** [review]
Thanks for your quick response on this PR! I think there's a few more missing `@experimental` tags and I asked for a some more clarification on the `index.test.ts` changes

File: src/collection.ts [hunk 7]
```diff
@@ -115,7 +115,7 @@ export interface CollectionOptions extends BSONSerializeOptions, WriteConcernOpt
   readConcern?: ReadConcernLike;
   /** The preferred read preference (ReadPreference.PRIMARY, ReadPreference.PRIMARY_PREFERRED, ReadPreference.SECONDARY, ReadPreference.SECONDARY_PREFERRED, ReadPreference.NEAREST). */
   readPreference?: ReadPreferenceLike;
-  /** @internal TODO(NODE-5688): make this public */
+  /** Specifies the time an operation will run until it throws a timeout error */
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/collection.ts:118]: done

File: src/cursor/abstract_cursor.ts [hunk 8]
```diff
@@ -61,15 +61,41 @@ export interface CursorStreamOptions {
...
+ */
 export const CursorTimeoutMode = Object.freeze({
   ITERATION: 'iteration',
   LIFETIME: 'cursorLifetime'
 } as const);
 
-/** @public
- * TODO(NODE-5688): Document and release
- * */
+/** @public */
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/cursor/abstract_cursor.ts:98]: done

File: src/db.ts [hunk 9]
```diff
@@ -97,7 +97,7 @@ export interface DbOptions extends BSONSerializeOptions, WriteConcernOptions {
   readConcern?: ReadConcern;
   /** Should retry failed writes */
   retryWrites?: boolean;
-  /** @internal TODO(NODE-5688): make this public */
+  /** Specifies the time an operation will run until it throws a timeout error */
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/db.ts:100]: done

File: src/gridfs/index.ts [hunk 10]
```diff
@@ -38,7 +38,10 @@ export interface GridFSBucketOptions extends WriteConcernOptions {
   chunkSizeBytes?: number;
   /** Read preference to be passed to read operations */
   readPreference?: ReadPreference;
-  /** @internal TODO(NODE-5688): make this public */
+  /**
+   * Specifies the lifetime duration of a gridFS stream. If any async operations are in progress
+   * when this timeout expires, the stream will throw a timeout error.
+   */
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/gridfs/index.ts:44]: done

File: src/mongo_client.ts [hunk 11]
```diff
@@ -130,7 +130,7 @@ export type SupportedNodeConnectionOptions = SupportedTLSConnectionOptions &
 export interface MongoClientOptions extends BSONSerializeOptions, SupportedNodeConnectionOptions {
   /** Specifies the name of the replica set, if the mongod is a member of a replica set. */
   replicaSet?: string;
-  /** @internal TODO(NODE-5688): This option is in development and currently has no behaviour.  */
+  /** Specifies the time an operation will run until it throws a timeout error */
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/mongo_client.ts:133]: done

File: src/operations/operation.ts [hunk 12]
```diff
@@ -35,7 +35,7 @@ export interface OperationOptions extends BSONSerializeOptions {
   /** @internal Hint to `executeOperation` to omit maxTimeMS */
   omitMaxTimeMS?: boolean;
 
-  /** @internal TODO(NODE-5688): make this public */
+  /** Specifies the time an operation will run until it throws a timeout error */
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/operations/operation.ts:38]: done

File: src/operations/run_command.ts [hunk 13]
```diff
@@ -15,7 +15,7 @@ export type RunCommandOptions = {
   session?: ClientSession;
   /** The read preference */
   readPreference?: ReadPreferenceLike;
-  /** @internal */
+  /** Specifies the time an operation will run until it throws a timeout error */
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/operations/run_command.ts:18]: done

File: src/sessions.ts [hunk 14]
```diff
@@ -145,7 +145,7 @@ export class ClientSession
   [kPinnedConnection]?: Connection;
   /** @internal */
   [kTxnNumberIncrement]: number;
-  /** @internal */
+  /** Specifies the time an operation in a given `ClientSession` will run until it throws a timeout error */
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/sessions.ts:148]: done

File: test/unit/index.test.ts (hunk 2 above)
Comment (nbbeeken) [reply test/unit/index.test.ts:8]: ts-node internally compiles the typescript before running the javascript, because we didn't explictily mark 'AnyBulkWriteOperation', 'BulkWriteOptions' as `type` so ts-node leaves behind empty exports with those names, our real JS in lib never had those, so this was a bit of cruft only related to how our tests run.

File: src/cursor/run_command_cursor.ts [hunk 15]
```diff
@@ -19,9 +19,41 @@ import {
 export type RunCursorCommandOptions = {
   readPreference?: ReadPreferenceLike;
   session?: ClientSession;
-  /** @internal */
+  /**
+   * Specifies the time an operation will run until it throws a timeout error. Note that if
+   * `maxTimeMS` is provided in the command in addition to setting `timeoutMS` in the options, then
+   * the original value of `maxTimeMS` will be overwritten.
+   */
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/cursor/run_command_cursor.ts]: done

**Review by aditi-khare-mongoDB — APPROVED** [review]
